File: examples/fidl2/echo2_client_cpp/echo2_client.cc
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <async/cpp/loop.h>
#include <async/default.h>
#include <launchpad/launchpad.h>
#include <zircon/processargs.h>
#include <zx/process.h>

#include "garnet/examples/fidl2/services/echo2.fidl.cc.h"
#include "lib/app/cpp/application_context.h"
#include "lib/fidl/cpp/binding_set.h"
#include "lib/svc/cpp/services.h"

namespace echo2 {

class EchoClientApp {
 public:
  EchoClientApp()
      : context_(app::ApplicationContext::CreateFromStartupInfo()) {}

  echo2::EchoPtr& echo() { return echo_; }

  void Start(std::string server_url, std::string msg) {
    auto launch_info = app::ApplicationLaunchInfo::New();
    launch_info->url = server_url;
    launch_info->service_request = echo_provider_.NewRequest();
    context_->launcher()->CreateApplication(std::move(launch_info),
                                            controller_.NewRequest());

    echo_provider_.ConnectToService(echo_.NewRequest().TakeChannel(),
                                    "echo2.Echo");
  }

 private:
  EchoClientApp(const EchoClientApp&) = delete;
  EchoClientApp& operator=(const EchoClientApp&) = delete;

  std::unique_ptr<app::ApplicationContext> context_;
  zx::process server_;
  app::Services echo_provider_;
  app::ApplicationControllerPtr controller_;
  echo2::EchoPtr echo_;
};

}  // namespace echo2

int main(int argc, const char** argv) {
  std::string server_url = "echo2_server_cpp";
  std::string msg = "hello world";

  for (int i = 1; i < argc - 1; ++i) {
    if (!strcmp("--server", argv[i])) {
      server_url = argv[++i];
    } else if (!strcmp("-m", argv[i])) {
      msg = argv[++i];
    }
  }

  async_loop_config_t config = {
      // The FIDL support lib requires async_get_default() to return non-null.
      .make_default_for_current_thread = true,
  };
  async::Loop loop(&config);

  echo2::EchoClientApp app;
  app.Start(server_url, msg);

  app.echo()->EchoString(msg, [](fidl::StringPtr value) {
    printf("***** Response: %s\n", value->data());
  });

  return app.echo().WaitForResponse();
}

File: examples/fidl/echo2_client_go/echo2_client.go
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package main

import (
	"flag"
	"fmt"

	"app/context"
	"svc/services"

	"fidl/component"
	"fidl/echo2"
)

type echoClientApp struct {
	ctx          *context.Context
	echoProvider *services.Provider
	controller   *component.ComponentControllerInterface
	echo         *echo2.EchoInterface
}

func (a *echoClientApp) startApplication(serverURL string) (li *component.ComponentControllerInterface, err error) {
	pr, err := a.echoProvider.NewRequest()
	if err != nil {
		return nil, fmt.Errorf("NewRequest failed: %v", err)
	}
	defer func() {
		if err != nil {
			pr.Close()
		}
	}()

	launchInfo := component.LaunchInfo{
		Url:              serverURL,
		DirectoryRequest: pr,
	}

	cr, cp, err := component.NewComponentControllerInterfaceRequest()
	if err != nil {
		return nil, fmt.Errorf("NewComponentControllerInterfaceRequest failed: %v", err)
	}
	defer func() {
		if err != nil {
			cr.Close()
			cp.Close()
		}
	}()

	err = a.ctx.Launcher.CreateApplication(launchInfo, cr)
	if err != nil {
		return nil, fmt.Errorf("CreateApplication failed: %v", err)
	}
	return cp, nil
}

func (a *echoClientApp) getEchoInterface() (ei *echo2.EchoInterface, err error) {
	r, p, err := echo2.NewEchoInterfaceRequest()
	if err != nil {
		return nil, fmt.Errorf("NewEchoInterfaceRequest failed: %v", err)
	}
	defer func() {
		if err != nil {
			r.Close()
			p.Close()
		}
	}()

	err = a.echoProvider.ConnectToService(r)
	if err != nil {
		return nil, fmt.Errorf("ConnectToServiceAt failed: %v", err)
	}
	return p, nil
}

func main() {
	serverURL := flag.String("server", "echo2_server_go", "server URL")
	msg := flag.String("m", "Hello, Go World", "message")

	flag.Parse()

	a := &echoClientApp{
		ctx:          context.CreateFromStartupInfo(),
		echoProvider: services.NewProvider(),
	}

	var err error
	a.controller, err = a.startApplication(*serverURL)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer a.controller.Close()

	a.echo, err = a.getEchoInterface()
	if err != nil {
		fmt.Println(err)
		return
	}
	defer a.echo.Close()

	response, err := a.echo.EchoString(msg)
	if err != nil {
		fmt.Println("EchoString failed:", err)
		return
	}

	fmt.Println("client:", *response)
}
